### ISSUE TITLE
Roundtrip is broken for Floats

### ISSUE THREAD

**Issue by Amval** [quinnj/JSON3.jl#39]
Hello, 

I observed this behaviour in your library and wanted to let you know. I am actually relying on a library that uses `JSON2`, which suffers from similar behaviour and it's causing some serialization issues for me. We were discussing the possibility of just changing the dependency to `JSON3`, but the problem persists.

```julia
using JSON
using JSON3

JSON.parse(JSON.json(2.0)) # Float -> Float
JSON3.read(JSON3.write(2.0)) # Float -> Int

# Dict. Although This might be intended?
JSON.parse(JSON.json(Dict("abc" => 1))) # Dict -> Dict
JSON3.read(JSON3.write(Dict("abc" => 1))) # Dict -> Object
```
I saw [something similar](https://github.com/quinnj/JSON2.jl/issues/10) reported in the `JSON2` repository.

Thank you for your time.

**Comment by quinnj** [comment]
I don't think there's much we can really do here when you're just relying on the "builtin" read machinery; this is just a natural part of the JSON spec, i.e. that a number can be `2` or `2.0`. In in this package, we try and be intelligent and map them to reasonable number types, but there's just not a way to reliably do what you want in all cases.

_Except_ that JSON3 also provides the really wonderful struct/typed reading as well; so if you really need a Float all the time, then just make sure you do `JSON3.read(json, Float64)`. Same with Dict, if you actually want a Dict, then do `JSON3.read(json, Dict)`.

Does that work?

**Comment by Amval** [comment]
It's currently not causing me any problems, as I made a pull request into the library I was using and replaced `JSON2` with `JSON` and it was accepted.

However, I do think the example I provided indicates that something is wrong. In particular, a smart autocasting of a `Float` into an `Int`.

See that the `JSON` library has the expected behaviour:
```julia
JSON.parse(JSON.json(2.0)) # => 2.0
JSON.parse(JSON.json(2) # => 2
```

But `JSON3`:
```julia
JSON3.read(JSON3.write(2.0)) # => 2
JSON.parse(JSON.json(2) # => 2
```

So the behaviour I would expect is possible, as shown by the `JSON` example. What does the struct/typed reading? Forces a type of casting overall?

Anyway, I have no need of this library changing anything, so if this behaviour is intentional is completely ok with me.
